lib/instrumentation/express-utils.js

```diff
@@ -57,7 +51,7 @@ function getPathFromRequest (req, useBase, usePathAsTransactionName) {
   }
 
   if (usePathAsTransactionName) {
-    const parsed = parseUrl(req)
+    const parsed = parsers.parseUrl(req)
```

Comment:
At 1st I thought to have separated APIs for string and request types but since the return types was exactly the same I decided to overload the method (not sure if this is allowed in your coding guidelines).

Based on your comment about leveraging `headers.host` to get the base for the URL resolution I tihnk the line would be as you described above

```
const parsed = (req.url.startsWith('/') ? new URL(base + req.url) : new URL(req.url, base))
```

First think that came to my mind was to provide the base as a param of the `parseUrl` function with a default value of `relative:///` but since you mentioned there is no comment about why the constant is there I prefer to leave it untouched.

Maybe my lack of context about what a `Transaction` is makes me wonder if customers that will get new behavior of the parsing logic, hence different transaction names, are going to be impacted somehow.  I guess if transaction is an entity that lives within a session is fine but not if the entity has bigger scope than the session.

If what is mentioned above is not an issue I'd go with the proposed changes:
- leave parsers untouched & remove tests (there is no point to test URL node API)
- leverage host header & base constructor param in `getPathFromRequest`
- add tests for `getPathFromRequest` method

Thanks for the thorough review @trentm and I wait for you feedback and 👍 to apply your suggestions.

Reply:
> Maybe my lack of context about what a `Transaction` is makes me wonder if customers that will get new behavior of the parsing logic, hence different transaction names

(If it helps, a "Transaction" is Elastic APM's name for a top-level tracing "span" in a service. In this case a Transaction is used to hold the tracing data for an incoming HTTP request. https://www.elastic.co/guide/en/apm/guide/current/data-model-transactions.html provides some details.)

I think adding tests for `getPathFromRequest` could cover all the same cases that "parseurl/test/test.js" covers for `parseurl(req)`. If there are any differences using `const parsed = (req.url.startsWith('/') ? new URL(base + req.url) : new URL(req.url, base))`, then we can discuss again. My guess is that all cases will give the same result with the newer implementation -- which means there will be no change in `transaction.name` behaviour for users.

Also, I think generally moving to the `URL` class for URL parsing and away from `url.parse`, `Url`, and the `parseurl` module is a good thing. https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost says

> url.parse() uses a lenient, non-standard algorithm for parsing URL strings. It is prone to security issues such as [host name spoofing](https://hackerone.com/reports/678487) and incorrect handling of usernames and passwords. Do not use with untrusted input. CVEs are not issued for url.parse() vulnerabilities. Use the [WHATWG URL](https://nodejs.org/api/url.html#the-whatwg-url-api) API instead.

I don't think the APM agent's usage of these is a concern for *security* issues, but the newer API is standardized and maintained.

> If what is mentioned above is not an issue I'd go with the proposed changes:
> ...

👍  Yes, it would be great if you could change your PR to do these.  One more thing is that once `parseurl` usage is removed you can apply this change for the `npm run test:deps` task:

```diff
diff --git a/package.json b/package.json
index 4ced2900..f96618b4 100644
--- a/package.json
+++ b/package.json
@@ -15,7 +15,7 @@
     "lint:yaml-files": "./dev-utils/lint-yaml-files.sh  # requires node >=10",
     "coverage": "COVERAGE=true ./test/script/run_tests.sh",
     "test": "./test/script/run_tests.sh",
-    "test:deps": "dependency-check index.js start.js start-next.js 'lib/**/*.js' 'test/**/*.js' '!test/activation-method/fixtures' '!test/instrumentation/modules/next/a-nextjs-app' --no-dev -i async_hooks -i perf_hooks -i parseurl -i node:http -i @azure/functions-core",
+    "test:deps": "dependency-check index.js start.js start-next.js 'lib/**/*.js' 'test/**/*.js' '!test/activation-method/fixtures' '!test/instrumentation/modules/next/a-nextjs-app' --no-dev -i async_hooks -i perf_hooks -i node:http -i @azure/functions-core",
     "test:tav": "tav --quiet && (cd test/instrumentation/modules/next/a-nextjs-app && tav --quiet)",
     "test:docs": "./test/script/docker/run_docs.sh",
     "test:types": "tsc --project test/types/tsconfig.json && tsc --project test/types/transpile/tsconfig.json && node test/types/transpile/index.js && tsc --project test/types/transpile-default/tsconfig.json && node test/types/transpile-default/index.js",
```

Reply:
Thanks @trentm for the clarification about transaction model :) 

Suggested changes are applied and pushed for you review